subprojects/logging/src/main/java/org/gradle/internal/logging/console/LogGroupingOutputEventListener.java

```diff
@@ -0,0 +1,229 @@
+        this.listener = listener;
+        this.executor = executor;
+        this.timeProvider = timeProvider;
+    }
+
+    @Override
+    public void onOutput(OutputEvent event) {
+        synchronized (lock) {
+            if (event instanceof EndOutputEvent) {
+                renderAllGroups(outputEventGroups);
+                listener.onOutput(event);
+//                executor.shutdown();
+            } else if (event instanceof ProgressStartEvent) {
+                onStart((ProgressStartEvent) event);
+            } else if (event instanceof ProgressEvent) {
+                ProgressEvent progressEvent = (ProgressEvent) event;
+                groupOrForward(progressIdToBuildOperationIdMap.get(progressEvent.getProgressOperationId()), progressEvent);
+            } else if (event instanceof RenderableOutputEvent) {
+                RenderableOutputEvent renderableOutputEvent = (RenderableOutputEvent) event;
+                groupOrForward(renderableOutputEvent.getBuildOperationId(), renderableOutputEvent);
+            } else if (event instanceof ProgressCompleteEvent) {
+                onComplete((ProgressCompleteEvent) event);
+            } else {
+                listener.onOutput(event);
+            }
+
+            // TODO(ew): Use strategy from ThrottlingOEL to schedule runnable once every 5 seconds on new output
+//            executor.schedule(new Runnable() {
+//                @Override
+//                public void run() {
+//                    synchronized (lock) {
+//                        renderAllGroups(outputEventGroups);
+//                    }
+//                }
+//            }, SCHEDULER_CHECK_PERIOD_MS, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private void onStart(ProgressStartEvent event) {
+        Object buildOpId = event.getBuildOperationId();
+        if (buildOpId != null) {
+            buildOperationIdForest.put(buildOpId, event.getParentBuildOperationId());
+            progressIdToBuildOperationIdMap.put(event.getProgressOperationId(), buildOpId);
+
+            if (event.getBuildOperationType() == BuildOperationType.TASK || event.getBuildOperationType() == BuildOperationType.CONFIGURE_PROJECT) {
+                // TODO: check whether child of another task
+                List<StyledTextOutputEvent.Span> header = Collections.singletonList(makeHeader(event));
+                CategorisedOutputEvent headerEvent = new StyledTextOutputEvent(event.getTimestamp(), event.getCategory(), LogLevel.QUIET, buildOpId, header);
+                outputEventGroups.put(buildOpId, Lists.newArrayList(headerEvent, event));
+            } else {
+                groupOrForward(buildOpId, event);
+            }
+        } else {
+            listener.onOutput(event);
+        }
+    }
+
+    private StyledTextOutputEvent.Span makeHeader(ProgressStartEvent event) {
+        final String message;
+        if (event.getLoggingHeader() != null) {
```

Comment:
The goals here are to make the display of these headers as close to other logs as possible, and give everyone flexibility to customize the logging headers if they so choose.

The logging header is currently used for task execution logging. Most, but not all, renderable events set a short description and this is what you see in the Console. 

Currently, the tasks and projects you see all use the `description` and logging header is ignored. That doesn't mean it'll stay that way.